msamp/optim/adamw.py

```diff
@@ -228,3 +229,105 @@ def adamw_fn(   # noqa: C901
+        tensor_scale=True,
+    ):
+        """Constructor. See LBAdamW class docstring for details."""
+        self.tensor_scale = tensor_scale
+        super().__init__(
+            params,
+            lr=lr,
+            bias_correction=bias_correction,
+            betas=betas,
+            eps=eps,
+            weight_decay=weight_decay,
+            amsgrad=False,
+            maximize=maximize,
+            exp_avg_dtype=exp_avg_dtype,
+            exp_avg_sq_dtype=exp_avg_sq_dtype
+        )
+
+        self.original_params = []
+        self.master_weights = []
+
+        for group in self.param_groups:
+            params = []
+            for param in group['params']:
+                if param is None:
+                    continue
+
+                self.original_params.append(param)
+                if hasattr(param, '_meta') and param._meta is not None and param.numel() > 0:
+                    dtype = Dtypes.qtype_to_dtype[param._meta.qtype]
+                    param = ScalingTensor(param.view(dtype), param._meta)
+                    master_weight = param.cast(Dtypes.kfloat16)
+                    master_weight.requires_grad = True
+                    self.master_weights.append(master_weight)
+                    params.append(master_weight)
+                else:
+                    self.master_weights.append(None)
+                    params.append(param)
+
+            group['params'] = params
+
+    def zero_grad(self, set_to_none=False):
+        """Zero gradients."""
+        for param in self.original_params:
+            if set_to_none:
+                param.grad = None
+            else:
+                if param.grad is not None:
+                    if param.grad.grad_fn is not None:
+                        param.grad.detach_()
+                    else:
+                        param.grad.requires_grad_(False)
+                    param.grad.zero_()
+
+    def step(self):
+        """Performs a single optimization step."""
+        # Set gradient of master weight.
+        for i, param in enumerate(self.original_params):
+            if self.master_weights[i] is not None:
+                grad_meta = param._grad_meta
+                dtype = Dtypes.qtype_to_dtype[grad_meta.qtype]
+                self.master_weights[i].grad = ScalingTensor(param.grad.view(dtype), grad_meta)
+                param.grad = None
+
+        # call step() to update master weight
+        super().step()
+
+        # Copy master weight to weight
+        for i, param in enumerate(self.original_params):
+            if hasattr(param, '_meta') and param._meta is not None:
+                hp_data = None
+                if param.numel() == 0:
+                    param._meta.amax[0].zero_()
+                else:
+                    hp_data = self.master_weights[i].float()
+                    param._meta.amax[0] = hp_data.abs().max()
+
+                dist.all_reduce(param._meta.amax[0], op=dist.ReduceOp.MAX)
```

Comment:
Why synchronize amax here?

`param` is shard in FSDP and stored only in a specific GPU.

I think it is unnecessary to synchronize `amax`.

Reply:
Just as synced offline, it is possible that one parameter is sharded to multiple GPUs, so it is necessary to synchronize `amax`.